### ISSUE TITLE
Question about infer with camera parameters

### ISSUE THREAD

**Issue by klan-security** [Livioni/OmniVGGT-official#9]
Dear author:
      
     I combined your code with native VGGT code and trained it on my own dataset. Why does including camera parameters during inference make the results worse. 
     I use images from different perspectives of the same scene and corresponding camera parameters for inference, but the point clouds obtained from different perspectives have a significant difference in z-values. When piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
     May I ask what could be the reason for this, the extrinsic i used is world2cam.

**Comment by klan-security** [comment]
Thans for your replay.

I replaced the aggregator in the native VGGT with the zeroaggregator you built, and also replaced the encoding method for pose with the one you proposed

**Comment by Livioni** [comment]
Hi, Here is our dirty code of the training loop:
```python
            for step, batch in enumerate(train_dataloader): 
                batch = merge_dicts(batch)
                # the batch['extrinsic'] is in world2cam format.
                new_extrinsics, _, new_world_points, new_depths = normalize_camera_extrinsics_and_points_batch(
                    extrinsics = batch['extrinsic'], 
                    cam_points = None, 
                    world_points = batch['world_points'], 
                    depths = batch['depth'],
                    point_masks = batch['valid_mask'],
                    )
            
                input_extrinsics = batch['extrinsic'].clone()
                input_depths = batch['depth'].clone()
                input_mask = batch['valid_mask'].clone()
                
                batch['extrinsic'] = new_extrinsics
                batch['world_points'] = new_world_points
                batch['depth'] = new_depths
                
      
                with torch.amp.autocast('cuda', dtype=weight_dtype):
                    inputs = {
                        'images': batch['images'],
                        'extrinsics' : input_extrinsics, #input the original extrinsics because we normalize them in the aggregator
                        'intrinsics' : batch['intrinsic'],
                        'depth' : input_depths,
                        'mask' : input_mask
                    }

                    predictions = model(**inputs)
                # loss is supposed to be symmetric
                loss_details = {}
                with torch.amp.autocast('cuda', enabled=False):
                    
                    loss_dict = train_criterion(predictions, batch) #same of VGGT
                    for key, value in loss_dict.items():
                        if isinstance(value, torch.Tensor):
                            loss_details[key] = value.detach().item()
                        else:
                            loss_details[key] = value

                accelerator.backward(loss_dict['objective'])
```

And we do stochastic training in the zeroaggregator. 
```python
        camera_gt_index = self.select_camera_gt(S)
        depth_gt_index  = self.select_depth_gt(S)
        
        if len(camera_gt_index) != 0:
            camera_gt_length = len(camera_gt_index)
            camera_idx_tensor = torch.tensor(camera_gt_index, device=depth.device)
            
            # extrinsics = input_extrinsics
            extrinsics_selected = torch.index_select(extrinsics, dim=1, index=camera_idx_tensor)
            intrinsics_selected = torch.index_select(intrinsics, dim=1, index=camera_idx_tensor)

            extrinsics_gt_normalized = self.normalize_extrinsics(extrinsics_selected)
            pose_encoding = extri_intri_to_pose_encoding(
                        extrinsics=extrinsics_gt_normalized,
                        intrinsics=intrinsics_selected,
                        image_size_hw=(H, W),
                        pose_encoding_type="absT_quaR_FoV",
            )
            gt_camera_token = self.pose_embeddings[0](pose_encoding).view(B * camera_gt_length, C).unsqueeze(1)
            
            device = depth.device
            camera_full = torch.zeros(K, 1, C, device=device, dtype=camera_token.dtype)

            camera_rows = (torch.arange(B, device=device).unsqueeze(1) * S + camera_idx_tensor.unsqueeze(0)).reshape(-1)
            camera_full[camera_rows] = gt_camera_token.to(dtype=camera_token.dtype)
            gt_camera_token = camera_full
        else:
            pose_encoding = None
            gt_camera_token = torch.zeros(K, 1, C, device=depth.device, dtype=camera_token.dtype)
  
        camera_token = camera_token + self.camera_adapters[0](gt_camera_token)

```
Hope this will help you.

**Comment by haofuly** [comment]
> Dear author:
> 
> ```
>  I combined your code with native VGGT code and trained it on my own dataset. Why does including camera parameters during inference make the results worse. 
>  I use images from different perspectives of the same scene and corresponding camera parameters for inference, but the point clouds obtained from different perspectives have a significant difference in z-values. When piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
>  May I ask what could be the reason for this, the extrinsic i used is world2cam.
> ```

@klan-security 
Hi, I meet the same problems: 
I use multi-view images together with their camera parameters. But when piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
Have you ever figure out where the bug lies?

**Comment by klan-security** [comment]
Hi,        I haven't pinpoint where the problem lies, but if conditions permit, you can try training a model that produces absolute scale point clouds directly. 

Through my experiments, this is feasible, and there is no layering phenomenon in point clouds from different perspectives







-----原始邮件-----
发件人:"Li Fuhao" ***@***.***>
发送时间:2025-12-31 16:19:00 (星期三)
收件人: Livioni/OmniVGGT-official ***@***.***>
抄送: klan ***@***.***>, Mention ***@***.***>
主题: Re: [Livioni/OmniVGGT-official] Question about infer with camera parameters (Issue #9)


haofuly left a comment (Livioni/OmniVGGT-official#9)

Dear author:

 I combined your code with native VGGT code and trained it on my own dataset. Why does including camera parameters during inference make the results worse. 
 I use images from different perspectives of the same scene and corresponding camera parameters for inference, but the point clouds obtained from different perspectives have a significant difference in z-values. When piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
 May I ask what could be the reason for this, the extrinsic i used is world2cam.


@klan-security
Hi, I meet the same problems:
I use multi-view images together with their camera parameters. But when piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
Have you ever figure out where the bug lies?

**Comment by Livioni** [comment]
> > Dear author:
> > ```
> >  I combined your code with native VGGT code and trained it on my own dataset. Why does including camera parameters during inference make the results worse. 
> >  I use images from different perspectives of the same scene and corresponding camera parameters for inference, but the point clouds obtained from different perspectives have a significant difference in z-values. When piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
> >  May I ask what could be the reason for this, the extrinsic i used is world2cam.
> > ```
> 
> [@klan-security](https://github.com/klan-security) Hi, I meet the same problems: I use multi-view images together with their camera parameters. But when piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon. Have you ever figure out where the bug lies?

Hi, @haofuly  thanks for the report — could you share your test case which shows the layering phenomenon so I can further check if there is anything wrong?

**Comment by choos11** [comment]
> Hi,        I haven't pinpoint where the problem lies, but if conditions permit, you can try training a model that produces absolute scale point clouds directly.
> 
> Through my experiments, this is feasible, and there is no layering phenomenon in point clouds from different perspectives
> […](#)
> 
> -----原始邮件-----
> 发件人:"Li Fuhao" ***@***.***>
> 发送时间:2025-12-31 16:19:00 (星期三)
> 收件人: Livioni/OmniVGGT-official ***@***.***>
> 抄送: klan ***@***.***>, Mention ***@***.***>
> 主题: Re: [Livioni/OmniVGGT-official] Question about infer with camera parameters (Issue [#9](https://github.com/Livioni/OmniVGGT-official/issues/9))
> 
> 
> haofuly left a comment ([Livioni/OmniVGGT-official#9](https://github.com/Livioni/OmniVGGT-official/issues/9))
> 
> Dear author:
> 
>  I combined your code with native VGGT code and trained it on my own dataset. Why does including camera parameters during inference make the results worse.
>  I use images from different perspectives of the same scene and corresponding camera parameters for inference, but the point clouds obtained from different perspectives have a significant difference in z-values. When piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
>  May I ask what could be the reason for this, the extrinsic i used is world2cam.
> 
> 
> [@klan-security](https://github.com/klan-security)
> Hi, I meet the same problems:
> I use multi-view images together with their camera parameters. But when piecing together point clouds from different perspectives, there will be a noticeable layering phenomenon.
> Have you ever figure out where the bug lies?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

Hi @klan-security, thanks for sharing that insight. You mentioned that "if conditions permit, you can try training a model that produces absolute scale point clouds directly. Through my experiments, this is feasible, and there is no layering phenomenon."
Could you please elaborate a bit on how to train a model for absolute scale point clouds directly? Any guidance on the specific steps or code modifications would be greatly appreciated. Thanks!

**Comment by klan-security** [comment]
Hi @choos11 , for train a model for absolute scale point clouds you can simply remove normalization during data processing
